netket/graph/lattice.py

```diff
@@ -261,6 +181,14 @@ def __init__(
                 to :math:`k`-nearest neighbor sites (measured by their Euclidean distance)
                 are included in the graph. The edges can be distiguished by their color,
                 which is set to :math:`k - 1` (so nearest-neighbor edges have color 0).
+                By default, nearest neighbours (:code:`max_neighbor_order=1`) are autogenerated
+                unless :code:`custom_edges` is passed.
+            custom_edges: Describes all edges for one unit cell that are distributed to
+                all unit cells. Should be a list of tuples, where each tuple contains
```

Comment:
sorry if I am late to the party: would it be possible to make this docstring a bit more explicit? for example I don't understand what `all edges for one unit cell that are distributed to all unit cells` means. Maybe also adding an example with this constructor in the docstring (down) might help?

Reply:
Maybe something like "`custom_edges` provides all edges starting within the unit cell, which are repeated in all copies of the unit cell in the constructed lattice." would help a bit (I think "distributed" sounds like a fixed amount of edges is redistributed, not that they are copied).

A small example (maybe just a manual non-Kitaev Honeycomb or so?) in the docstring could be nice too, I agree.

Reply:
sorry I am still very much confused by the docstring for custom_edges, here you say 
```Should be a list of tuples, where each tuple contains the indices of the starting and ending sites in the unit cell, it  vector length, and an optional color.``` but in the kitaev example you pass a vector, not a vector length ?

```python
        custom_edges=[
            (0, 1, [0.5, 0.5 / 3 ** 0.5]),
            (0, 1, [-0.5, 0.5 / 3 ** 0.5]),
            (0, 1, [0, -1 / 3 ** 0.5]),
        ],
```

maybe it's just that I don't understand exactly how this is setting 3 edges between two sites

Reply:
The current docstring contains this text:
```
Should be a list of tuples; each tuple should contain the following:
                    index of the starting point in the unit cell
                    index of the endpoint in the unit cell
                    vector distance between the two
                    color of the edge (optional)
```
Is "vector distance" better?

Reply:
I guess most of the confusion (for me) stems from ```vector length``` that should be ```vector difference``` instead or something similar... is it checked in the code that the provided vector difference is compatible with the lattice points? ( I guess so, but I didn't dig too deeply to understand where)

Reply:
> Is "vector distance" better?

not really, a distance is by definition a scalar.

Reply:
also, here it should be sufficient to specify the direction only, not sure why we need to pass a non-unit vector, but maybe I am missing something...

Reply:
> also, here it should be sufficient to specify the direction only, not sure why we need to pass a non-unit vector, but maybe I am missing something...

On a chain of form 0-1-2-3-4, starting from site 0 all sites are in the same direction, so we do need the full difference vector, not just the direction, unless I'm missing something.

Reply:
the signed vector direction

Reply:
> > also, here it should be sufficient to specify the direction only, not sure why we need to pass a non-unit vector, but maybe I am missing something...
> 
> On a chain of form 0-1-2-3-4, starting from site 0 all sites are in the same direction, so we do need the full difference vector, not just the direction, unless I'm missing something.

Exactly, and of course it comes up on higher dimensional lattices too.

Reply:
? sorry but the difference here can be -1 or +1 I don't see the problem :) or do you have edges across different unit cells?

Reply:
Yeah, there's no reason to stop someone from defining a second-neighbour interaction on a chain or something like that

Reply:
ok ok I see... I just hope we have tests for all these cases ehhe

Reply:
I mean, we couldn't construct any connected lattice if we couldn't move to other unit cells?

Reply:
no I meant next-to-nearest connections, but I see now that it's fine... I agree. What I mean is checking that the vector distance is compatible with the underlying lattice unit-cell vectors, for example you might input a distance vector that goes 30 degrees north-east in a square lattice... or something like that

Reply:
Ah I see, but that's not a problem because we require the exact vector between the two, and check whether it corresponds to a valid displacement between two sites in those sublattices, i.e. whether `distance + site_offsets[sl1] - site_offsets[sl2]` is a lattice vector:
https://github.com/netket/netket/blob/53ad8c1c212d0b161f969cff7e32a2acb048dd51/netket/graph/_lattice_edge_logic.py#L141-L143
So your scenario would definitely error out

Reply:
The code does check whether the specified edges actually connect two sites (with the specified unit cell indices), so e.g. this incorrect 45 deg edge in a square does throw an error:
```python
>>> nk.graph.Lattice([[1, 0], [0, 1]], [4, 4], custom_edges=[(0, 0, [0.5, 0.5])])
ValueError: [0.5, 0.5] is invalid distance vector between sublattices 0->0
```
When an actual site exists at the target position, however, everything works, so if someone wants the diagonal edges in a square, that is possible:
```python
>>> nk.graph.Lattice([[1, 0], [0, 1]], [4, 4], custom_edges=[(0, 0, [1, 1])])
# succeeds
```

Reply:
Ok amazing, that was I was looking for thanks! To be really pedantic, the error message should contain vector difference or so and not vector distance but it's minor aha